### DISCUSSION TITLE
Auto generating typescript types using hooks & pocketbase-typegen

### DISCUSSION THREAD

**Discussion by silenceofthewans** [pocketbase/pocketbase#1838] [Show and tell]
I'm new to GO and coming from tRPC and prisma, I really miss the automatic types but for now this is how I'm using [pocketbase-typegen](https://github.com/patmood/pocketbase-typegen), [collections hooks](https://pocketbase.io/docs/event-hooks/#collection-api-hooks) and a [cobra command](https://github.com/spf13/cobra) to introspect sqlite db to generate typescript types everytime I update my collection's fields. It's useful when I'm prototyping but still want to be typesafe.

Basically, I register a custom command using cobra which runs pocketbase-typegen, I then invoke this custom command in the relevant collections hooks. In the following code, I'm using [just](https://github.com/casey/just) to handle the actual running of pocketbase-typegen but you can directly use the command here. (_I keep all my commands in the root directory of project in a justfile_)

Cobra Command (replace var command with your pocketbase-typegen command):

```GO
package types

import (
	"fmt"
	"log"
	"os/exec"
	"strings"

	"github.com/spf13/cobra"
)

func GenerateTypescriptTypes() *cobra.Command {
	var generateTypes = &cobra.Command{
		Use:   "ts-types",
		Short: "generate types",
		Long:  "This command will generate types",
		Run: func(cmd *cobra.Command, args []string) {
			fmt.Println("generating types")
			var command = "just types" //Generates typscript types using `just` & `pocketbase-typegen`
			parts := strings.Fields(command)
			execCommand := exec.Command(parts[0], parts[1:]...)
			out, err := execCommand.Output()
			if err != nil {
				log.Fatal(err)
			} else {
				log.Printf("%s", out)
			}
		},
	}
	return generateTypes
}

```

Then I call this in collection hooks:

```GO
package hooks

import (
	"os"

	"github.com/username/repo/backend/utils/types" // location of GenerateTypescriptTypes()
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
)

func generateTypes() {
	types.GenerateTypescriptTypes().Run(nil, nil)
}

func Collections(app *pocketbase.PocketBase) {

	env := os.Getenv("development")

	if env == "development" {
		app.OnCollectionAfterCreateRequest().Add(func(e *core.CollectionCreateEvent) error {
			generateTypes()
			return nil
		})

		app.OnCollectionAfterUpdateRequest().Add(func(e *core.CollectionUpdateEvent) error {
			generateTypes()
			return nil
		})

		app.OnCollectionAfterDeleteRequest().Add(func(e *core.CollectionDeleteEvent) error {
			generateTypes()
			return nil
		})

		app.OnCollectionsAfterImportRequest().Add(func(e *core.CollectionsImportEvent) error {
			generateTypes()
			return nil
		})
	}

}

```

Additionally, I also add the command to RootCmd so I can directly run the command e.g. `pocketbase ts-types`

```GO
package hooks

import (
	"github.com/username/repo/backend/utils/types"
	"github.com/pocketbase/pocketbase"
)

func Commands(app *pocketbase.PocketBase) {
	app.RootCmd.AddCommand(types.GenerateTypescriptTypes())
}
```

Eventually, I would love for pocketbase's js-sdk to handle this and be db aware. Something like a `pocketbase generate` command which can do codegen like [prisma client](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/generating-prisma-client), so we can access collections like `pocketbase.myCollectionWhichExists.getList(1,20)` and get intellisense based on the db schema.

Feel free to suggest improvements to above code :)

**Comment by Dastari** [comment]
This is really neat. I also came from **Prisma** but using **TypeGraphQL** and was using **graphql-codegen** to generate types. Since moving to Pocketbase I've been looking for a similar way to achieve this.

**Comment by einarpersson** [comment]
Is this possible from javascript hooks now with 0.17?

Since `goja` is somewhat I guess I cannot use child_process to run `npx pocketbase-typegen ....`. 
=> Is it possible to somehow run a shell command from hooks?

I also tried `const typegen = require('pocketbase-typegen')` but no luck
```
2023/08/25 07:52:23 GoError: SyntaxError: node_modules/pocketbase-typegen/dist/index.js: Line 1:38 Unexpected token ILLEGAL (and 33 more errors) at github.com/dop251/goja_nodejs/require.(*RequireModule).require-fm (native)
```

**Reply by ganigeorgiev** [reply]
In order the `require('pocketbase-typegen')` to work you need to have a `node_modules` directory with the provided module (_note the [modules caveat in the docs](https://pocketbase.io/docs/js-overview/#loading-modules)_).
But I'm not sure that I understand what you are trying to do.
Keep in mind that the goja vm is not a TypeScript runtime and if you want to use TypeScript you'll need to transpile it on your own to plain JS files.
If you just want to use the TS declarations for autocompletion, you can "load" them in JS with the [triple-slash `reference` directive](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html).

> Is it possible to somehow run a shell command from hooks?

Yes, see [$os.exec()](https://pocketbase.io/jsvm/functions/_os.exec.html).

**Reply by einarpersson** [reply]
Thank you @ganigeorgiev once again for such quick reply, I think this solves it!

I am simply trying (as the original poster) to generate typescript types (using [pocketbase-typegen](https://github.com/patmood/pocketbase-typegen)) from the schema, and doing so automatically whenever something is changed.

In `package.json`:
```json
"db:typegen": "pocketbase-typegen --db ./db/pb_data/data.db --out ./src/lib/pocketbase-types.ts"
```

In `db/pb_hooks/typegen.pb.js`:
```js
$os.exec('npm', 'run', 'db:typegen').run()
```

**Reply by einarpersson** [reply]
@ganigeorgiev 

Should I use `onModelAfter____` or `onCollectionAfter____` for triggering type generation? I tried with `onModelAfter___` and it seems to work, but @silenceofthewans uses `onCollectionAfter___`. Does it matter?

Once I get this working, maybe I can add it somewhere to the docs as an example? I cannot be the only one who would think it is great to just get auto typegen working.

**Reply by ganigeorgiev** [reply]
@einarpersson

The `*Request` hooks are triggered **only** when the related web API action is performed. The `onModel*` hooks are triggered for every db model operation performed by the `app.Dao()` (records, collections, admins, etc.).

In general the `onModel*` hooks are more universal since they will be triggered even when the collection change was done programmatically in code (eg. `app.Dao().SaveCollection(collection)`), but you need to manually exclude the collection model type if you decide to use it. When using Go you can rely on the `*models.Collection` type assertion result to filter only the collection models, but in the goja hooks this cannot be used directly and you have to seek some other mechanism to filter it (eg. check if `e.model.tableName()` is `_collections`).

**Reply by einarpersson** [reply]
I did not understand the `onModel*`-hook completely (would it run even when SaveCollections is called in migrations? Does not seem like it)

I chose `onCollection` and ended up with the following:

```js
/// <reference path="../pb_data/types.d.ts" />

onCollectionAfterCreateRequest(({ collection }) => {
	if (!collection) return

	console.log(`Collection ${collection.name} created - generating types...`)
	$os.exec('npm', 'run', 'db:typegen').run()
})

onCollectionAfterUpdateRequest(({ collection }) => {
	if (!collection) return

	console.log(`Collection ${collection.name} updated - generating types...`)
	$os.exec('npm', 'run', 'db:typegen').run()
})

onCollectionAfterDeleteRequest(({ collection }) => {
	if (!collection) return

	console.log(`Collection ${collection.name} deleted - generating types...`)
	$os.exec('npm', 'run', 'db:typegen').run()
})
```

(with `"db:typegen": "pocketbase-typegen --db ./db/pb_data/data.db --out ./src/lib/pocketbase-types.ts",` in `package.json`)

Works well, so far!

**Reply by ganigeorgiev** [reply]
> would it run even when SaveCollections is called in migrations?

No, because inside the migrations you don't use `app.Dao()` but a locally created `Dao` instance (see also https://pocketbase.io/docs/go-database/#dao-without-event-hooks or https://pocketbase.io/docs/js-database/#dao-without-event-hooks)

**Reply by Zehir** [reply]
Work great but have an issue when you have also a watcher that listen for `./src/lib/pocketbase-types.ts` and then compile back `pb_hooks` files that restart back pocketbase and make the saving failed.

As a workaround I made a dalayed command, 1 second is enought to let pocketbase save the collection:
```json
    "typegen:pocketbase:delayed": "sleep 1 && pnpm run typegen:pocketbase >>/dev/null &",
```
